truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java

```diff
@@ -0,0 +1,216 @@
+        }
+
+        public static SetThreadLocalInFrameStrategy ofBlock(RubyContext context, DynamicObject block, String varName) {
+            Frame frame = Layouts.PROC.getDeclarationFrame(block);
+            if (frame == null) {
+                return NullStrategy.INSTANCE;
+            }
+            return of(context, frame, varName);
+        }
+
+        public static SetThreadLocalInFrameStrategy of(RubyContext context, Frame aFrame, String variableName) {
+            MaterializedFrame callerFrame = aFrame.materialize();
+            FrameDescriptor fd = callerFrame.getFrameDescriptor();
+
+            int depth = getVariableDeclarationFrameDepth(callerFrame, variableName);
+            MaterializedFrame mf = RubyArguments.getDeclarationFrame(callerFrame, depth);
+            FrameSlot fs = getVariableFrameSlotWrite(mf, variableName);
+            Object defaultValue = mf.getFrameDescriptor().getDefaultValue();
+            if (depth != 0) {
+                return new ComplexStrategy(fd, fs, defaultValue, depth);
+            } else {
+                return new SimpleStrategy(fd, fs, defaultValue);
+            }
+        }
+    }
+
+    private static class NullStrategy extends SetThreadLocalInFrameStrategy {
+        private static NullStrategy INSTANCE = new NullStrategy();
+
+        private NullStrategy() {
+            super(null, null, null);
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame) {
+            return null;
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocalBlock(RubyContext context, DynamicObject block) {
+            return null;
+        }
+
+        @Override
+        public boolean matches(Frame callerFrame) {
+            return callerFrame == null;
+        }
+    }
+
+    private static class SimpleStrategy extends SetThreadLocalInFrameStrategy {
+        public SimpleStrategy(FrameDescriptor fd, FrameSlot fs, Object defaultValue) {
+            super(fd, fs, defaultValue);
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame) {
+            return getThreadLocalObjectFromFrame(context, callerFrame, fs, defaultValue, true);
+        }
+    }
+
+    private static class ComplexStrategy extends SetThreadLocalInFrameStrategy {
+        private final int depth;
+
+        public ComplexStrategy(FrameDescriptor fd, FrameSlot fs, Object defaultValue, int depth) {
+            super(fd, fs, defaultValue);
+            this.depth = depth;
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame) {
+            final MaterializedFrame frame = RubyArguments.getDeclarationFrame(callerFrame, depth);
+            return getThreadLocalObjectFromFrame(context, frame, fs, defaultValue, true);
+        }
+    }
```

Comment:
I don't think there is any need to separate these code paths.
SimpleStrategy is just ComplexStrategy with depth=0 and
`RubyArguments.getDeclarationFrame(frame, depth);` supports depth=0 and optimizes just as good.